std/assembly/typedarray.ts

```diff
@@ -1718,3 +1762,70 @@ function WRAP<TArray extends ArrayBufferView, T>(buffer: ArrayBuffer, byteOffset
   store<usize>(out, changetype<usize>(buffer) + <usize>byteOffset, offsetof<TArray>("dataStart"));
   return changetype<TArray>(out); // retains
 }
+
+// @ts-ignore: decorator
+@inline
+function SET<TArray extends ArrayBufferView, T, UArray extends ArrayBufferView, U>(target: TArray, source: UArray, offset: i32 = 0): void {
+  // need to assert at compile time that U is not a reference or a function
+  if (isReference<U>()) {
+    ERROR(E_NOTIMPLEMENTED);
+  }
+
+  // Uncaught RangeError: offset is out of bounds
+  if (offset < 0) throw new RangeError(E_INDEXOUTOFRANGE);
+  if (source.length + offset > target.length) throw new RangeError(E_INDEXOUTOFRANGE);
+
+  // if the types align and match, use memory.copy() instead of manual loop
+  if (isInteger<T>() == isInteger<U>() && alignof<T>() == alignof<U>()) {
+    memory.copy(
+      target.dataStart + (<usize>offset << alignof<T>()),
+      source.dataStart,
+      source.byteLength
+    );
+  } else {
+    let targetDataStart = target.dataStart + (<usize>offset << alignof<T>());
+    let sourceDataStart = source.dataStart;
+    let count = source.length;
+    for (let i = 0; i < count; i++) {
+      // if TArray is Uint8ClampedArray, then values must be clamped
+      if (target instanceof Uint8ClampedArray) {
+        if (isFloat<U>()) {
+          let value = load<U>(sourceDataStart + (<usize>i << alignof<U>()));
+          store<T>(
+            targetDataStart + (<usize>i << alignof<T>()),
+            isFinite<U>(value) ? <T>max<U>(0, min<U>(255, value)) : 0
+          );
+        } else {
+          let value = load<U>(sourceDataStart + (<usize>i << alignof<U>()));
```

Comment:
Wondering if we'll get more sign-extension/masking here than necessary, and if so whether this can be avoided by typing the intermediate local `i/u32`. Like, if `value` is typed `u16` and we cast that to `i8`, the explicit conversion will wrap, even though the `store8` would already wrap.

Reply:
You mean `store<u8>(<u8>~( .... ))` is unnecessary and just need `load<u8>(~(... ))`?

Reply:
Mostly relevant if `T` and `U` have different sizes, which Wasm has distinct instructions for when loading and storing. The `store<T>` builtin is more of a `store<T>(<U>...)`. It evaluates the argument type in order to pick the right instruction, which implicitly wraps the value. So, if we explicitly convert `<T>someU` that might already be implicit in the store, and any explicit wrapping being unnecessary. Didn't look at the fixtures yet.

Reply:
Happy to make changes for that. Just want to make sure I understand it before I pull changes from master and make a qualitative change to the code.

Can you help me understand this?

Reply:
I believe what happens in the `T=i8, U=i16` case is that on `value = load<i16>(...), the compiler remembers that `value` is `i16` and wrapped (because the load returns a properly wrapped value). Then, `min<i16>(some, thing)` requires properly wrapped values to determine the minimum, and fortunately it's already wrapped so nothing to do there. After that, the value is cast to `<T>` (`i8` here), which wraps the i16 to an i8 discarding garbage bits (explicit cast does this), but that's unnecessary since what follows is a `store<i8>(...)` translating into an `i32.store8` that already discards garbage bits.

One first thing to test can be to simply remove the `<T>` at the start of the store argument, and see what happens.

Reply:
![image](https://user-images.githubusercontent.com/3761339/71109111-91022480-2192-11ea-8133-b13a8323c5e5.png)

Wow. @dcodeIO that was an incredible find. A typecast causing an optimization problem like that seems so obscure, I can't believe it.

Anyway, this should be fixed and brought current with master!

Reply:
I removed the offending `<T>`, and this should at least be resolved from an implementation standpoint, if not from the testing standpoint. See above!

Can we mark this conversation as resolved?